### ISSUE TITLE
symbol should not contain svg

### ISSUE THREAD

**Issue by TxHawks** [edenspiekermann/sprite.sh#4]
Cool project - the more workflows we have to easily produce svg sprites, the better.

I have a few suggestions for improvement:
1. It would be good if `sprite.sh` wouldn't force all icon to have the same `viewbox`, but rather if by default if checked for the original `SVG`'s `viewbox` and set it on the `symbol`. There could be an option for normalizing the `viewbox`, but I don't think it should be the default as there are many use cases in which it is counter productive (E.g., when different icons have different aspect ratios).
2. I think it is somewhat dangerous to automatically infer a `<title>` from the file name, as asstive technology may read it out loud. Also, I'm not sure how useful it is, as many screen readers don't automatically read the title, but rather depend on the `aria-labeledby` attribute. The generated `<title>` tag can not be refereed, since it has no `id`. 
3. I think it would be better if the generated [sprite.svg](https://raw.githubusercontent.com/edenspiekermann/sprite.sh/master/example/sprite.svg), did not contain the original svg inside it, as there is simply no need for it, and it adds extra weight to the generated sprite.  

to sum up, I would suggest that the following svg

``` svg
<!-- brand.svg -->
<svg version="1.1" 
         id="Ebene_1" 
         xmlns="http://www.w3.org/2000/svg" 
         xmlns:xlink="http://www.w3.org/1999/xlink" 
         x="0px" y="0px"
         width="20px" height="20px" 
         viewBox="0 0 20 30" 
         enable-background="new 0 0 20 30" 
         xml:space="preserve">
  <!-- content -->
</svg>
```

Would be transformed into: 

``` svg
<svg xmlns='http://www.w3.org/2000/svg' style='display: none;'>
  <symbol id='icon-brand' viewBox="0 0 20 30">
      <!-- Content -->
  </symbol>
 <!-- Other symbols -->
</svg>
```

This saves quite a lot of redundant code, and could end up pretty meaningful when generating a sprite with many different icons.

**Comment by TxHawks** [comment]
I realize these are actually three separate different issues, so if it's more convinient, I can split them into three Github issues

**Comment by KittyGiraudel** [comment]
Hello. Thanks for all the interesting suggestions! Let me go through them in order.
1. This is something I should be able to do. If the viewbox parameter is defined, use it or do not set one at all. Or set a default one. Not sure yet. But I can probably do something like this.
2. Maybe we could have a `--title` option that would do just that. That sounds like a fair requirement. I’ll consider it.
3. I’d argue this is not spritesh’ role to perform SVG optimisation. However I added [some recommendations to the README](https://github.com/edenspiekermann/sprite.sh#svg-optimisation) about how to optimise the initial SVG files to allow spritesh to perform at its best.

**Comment by TxHawks** [comment]
Thanks for the quick reply.

:thumbsup: re no. 1 & 2. A --title option sounds great. Also, if a title is generated, it's a good idea to give it an `id` attribute, so that it can be referred to by `aria-labelledby`.

Regarding no. 3, allow me to clarify - the original SVG must contain an `svg` tag, obviously, and some other attributes, notably `xmlns`. The user really just _cannot_ safely omit them.

However, when concatenating the individual SVGs into a sprite, the tag and the attributes become unnecessary.

While I generally agree with you that it isn't the script's role to do general svg optimizations on the SVGs passed to it, I do think that it should optimize its own output. This is the kind of optimization an svg optimizer could not later perform (it has no way of asserting whether or not the inner `svg` tag is needed or not).

Since according to your write up, the script is mostly meant to be used in situations where you don't have a full fledged build process, I think it would also be good if it didn't _insert_ whitespace (e.g. between `symbol` tags). Again, optimizing its own output, not that of the SVGs passed to it.

**Comment by KittyGiraudel** [comment]
1. is tackled in #6.
2. is tackled in #7. The `--title` option might not be necessary.
3. has been partially done in https://github.com/edenspiekermann/sprite.sh/commit/f2abf594c103e224ba32e50d07384a33ba1ebde7. The `<svg>` stripping itself cannot be done as it would require a HTML parser which cannot be done in Bash.

**Comment by HugoGiraudel** [comment]
Regarding 2., which element should have `aria-labelledBy`? The `svg`?

**Comment by TxHawks** [comment]
You would use `aria-labelledby` on the `svg` using the icon, like this:

``` svg
<!-- sprite -->
<svg xmlns='http://www.w3.org/2000/svg' style='display: none;'>
  <symbol id='icon-brand' viewBox="0 0 20 30">
    <title id='icon-brand__title'>edenspiekermann</title>
      <!-- Content -->
  </symbol>
 <!-- Other symbols -->
</svg>
```

``` html
<!-- index.html -->
<svg class='logo' aria-labelledby="icon-brand__title" >
    <use xlink:href="#icon-brand"></use>
</svg>
```

**Comment by TxHawks** [comment]
Re the change in #7, Say you have your logo in `logo.svg`, which is a completely reasonable choice, its automatically generated title would be `logo`, meaning that when someone accesses the page using assistive technology, it will read `logo`, where you would, in fact, reasonablly want it to read `edenspiekermann`.

Also, I'm not sure if setting the `aria-labelledby` on the symbol will actually work. Therre is the whole shadow-DOM barrier thing there, and I'm not sure if a screen reader knows how to cross that.

`aria-labelledby` should probably be placed on the `svg` that has the `use` tag, and be at the responsibility of the user.

[Using ARIA to enhance SVG accessibility](https://www.paciellogroup.com/blog/2013/12/using-aria-enhance-svg-accessibility) is a good resource on this

**Comment by KittyGiraudel** [comment]
Gotcha. I’ll rework this PR today.

**Comment by KittyGiraudel** [comment]
> Re the change in #7, Say you have your logo in logo.svg, which is a completely reasonable choice, its automatically generated title would be logo, meaning that when someone accesses the page using assistive technology, it will read logo, where you would, in fact, reasonablly want it to read edenspiekermann.

How would you tackle this then?

**Comment by TxHawks** [comment]
I would say that `<title>` and `aria-labelledby` should be within the responsibility of the user. If a user includes a `<title>` tag in their icon, it will be included in the `<symbol>`, and they should give it an `id`.

Since `aria-labelledby` should live on the using SVG, not in the sprite, it should definitely be within the responsibility of the user.

**Comment by KittyGiraudel** [comment]
So remove the `title` altogether? Sounds good to me.

**Comment by TxHawks** [comment]
:+1: 

The auto-generated `<title>` that is, right? If someone has a title in their original SVG, it would not be stripped from the symbol, correct?

**Comment by KittyGiraudel** [comment]
Exactly.

**Comment by KittyGiraudel** [comment]
1 and 2 are done. 3 is impossible. I’m closing, thanks! :)
